wdl-grammar/src/v1/lint/newline_eof.rs

```diff
@@ -0,0 +1,168 @@
+//! WDL files must end with a newline
+
+use std::collections::VecDeque;
+
+use nonempty::NonEmpty;
+use pest::iterators::Pair;
+use wdl_core::concern::code;
+use wdl_core::concern::lint;
+use wdl_core::concern::lint::Group;
+use wdl_core::concern::lint::Rule;
+use wdl_core::concern::Code;
+use wdl_core::file::Location;
+use wdl_core::Version;
+
+use crate::v1;
+
+/// Detects incorrect count of newline at EOF
+#[derive(Debug)]
+pub struct NewlineEOF;
+
+impl<'a> NewlineEOF {
+    /// Creates a warning for a file ending with a newline
+    fn missing_newline_at_eof(&self, location: Location) -> lint::Warning
+    where
+        Self: Rule<&'a Pair<'a, v1::Rule>>,
+    {
+        // SAFETY: this error is written so that it will always unwrap.
+        lint::warning::Builder::default()
+            .code(self.code())
+            .level(lint::Level::Low)
+            .group(self.group())
+            .push_location(location)
+            .subject("missing newline at the end of the file")
+            .body("There should always be a newline at the end of a WDL file.")
+            .fix("Add a newline at the end of the file.")
+            .try_build()
+            .unwrap()
+    }
+
+    /// Creates a warning for a file ending with more than one newline
+    fn no_empty_line_at_eof(&self, location: Location) -> lint::Warning
+    where
+        Self: Rule<&'a Pair<'a, v1::Rule>>,
+    {
+        // SAFETY: this error is written so that it will always unwrap.
+        lint::warning::Builder::default()
+            .code(self.code())
+            .level(lint::Level::Low)
+            .group(self.group())
+            .push_location(location)
+            .subject("erroneous empty line at the end of file")
+            .body("There should only be a newline at the end of a WDL file.")
+            .fix("Remove the empty line at the end of the file.")
+            .try_build()
+            .unwrap()
+    }
+}
+
+impl<'a> Rule<&'a Pair<'a, v1::Rule>> for NewlineEOF {
+    fn code(&self) -> Code {
+        // SAFETY: this manually crafted to unwrap successfully every time.
+        Code::try_new(code::Kind::Warning, Version::V1, 7).unwrap()
+    }
+
+    fn group(&self) -> Group {
+        Group::Spacing
+    }
+
+    fn check(&self, tree: &'a Pair<'_, v1::Rule>) -> lint::Result {
+        let mut warnings: VecDeque<_> = VecDeque::new();
+
+        let mut iter = tree
```

Comment:
This seems precarious. The way I would approach this in a vacuum is (a) call `into_iter()` but don't call `flatten()` afterwards and then (b) look at the end of the list for how many newlines are present. I think an approach like that will be more stable as the tree continues to evolve (maybe we will break the assumptions that lines 78 and 79 make in the future).